src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -112,6 +104,98 @@ int IndexOf(string str, char c)
             }
         }
 
+        private static bool TryGetNormalizedVersion(string str, [NotNullWhen(true)] out Version? version)
+        {
+            if (!string.IsNullOrEmpty(str) && GetNextSection(str, 0, out int endIndex, out int major))
+            {
+                int build = 0;
+                int revision = 0;
+
+                // check for all the possible parts of the version string. If endIndex is less than the end of
+                // the string, the input string was invalid (e.g. "1.2.3.4.5").
+                bool validString = GetNextSection(str, endIndex + 1, out endIndex, out int minor) &&
+                    GetNextSection(str, endIndex + 1, out endIndex, out build) &&
+                    GetNextSection(str, endIndex + 1, out endIndex, out revision) &&
+                    endIndex == str.Length;
+
+                if (validString)
+                {
+                    version = new Version(major, Math.Max(0, minor), Math.Max(0, build), Math.Max(0, revision));
+                    return true;
+                }
+            }
+
+            version = null;
+
+            return false;
+
+            // returns false if an invalid section was found while processing the string
+            static bool GetNextSection(string s, int start, out int end, out int versionNumber)
+            {
+                // check to see if we've processed the whole string
+                if (start >= s.Length)
+                {
+                    // we've reached the end. The section is empty but not invalid.
+                    end = s.Length;
+                    versionNumber = -1;
+
+                    return true;
+                }
+
+                end = s.IndexOf('.', start);
+
+                if (end == -1)
+                {
+                    end = s.Length;
+                }
+
+                return TryParseInt(s, start, end, out versionNumber);
+            }
+
+            // start is inclusive bound and end is exclusive for the section of string to parse
+            static bool TryParseInt(string s, int start, int end, out int value)
+            {
+                // invalid section specified
+                if (start >= end)
+                {
+                    value = 0;
+                    return false;
+                }
+
+                long returnValue = 0;
+
+                // Parse the values in reverse and assign a tens place to them.
+                // When parsing "123456", this method adds 6 + 50 + 400 + 3000 + 20000 + 100000
+                for (int i = start; i < end; i++)
+                {
+                    // negative numbers are invalid for version strings so we only need to check for digits
+                    char current = s[i];
+                    if (current < '0' || current > '9')
```

Comment:
One more edge case here.  For values with spaces in them, `Version.TryParse()` appears to handle this gracefully.  This new code does not.

For example, if my version is `1. 2`, the new code will fail to parse because of the space.  (https://dotnetfiddle.net/i7hx5x)

It appears to only ignore leading and trailing whitespace though.  I'd hate to call `.Trim()` but do we want to have things work if the user specifies:

```xml
<PackageReference Include="PackageA" Version="1. 23 . 4" />
```

Note, our code works with that crazy version today...